contracts/0.8.9/WithdrawlQueueBase.sol

```diff
@@ -0,0 +1,344 @@
+     */
+    function claimWithdrawal(uint256 _requestId, uint256 _hint) external {
+        if (_requestId == 0) revert ZeroRequestId();
+        if (_requestId > lastFinalizedRequestId) revert RequestNotFinalized(_requestId);
+
+        WithdrawalRequest storage request = queue[_requestId];
+        if (request.claimed) revert RequestAlreadyClaimed();
+
+        request.claimed = true;
+
+        Discount memory discount;
+        if (_hint <= lastDiscountIndex && _compareWithHint(_requestId, _hint) == 0) {
+            discount = discountHistory[_hint];
+        } else {
+            revert InvalidHint(_hint);
+        }
+
+        uint128 ethToSend = queue[_requestId].cumulativeStETH - queue[_requestId - 1].cumulativeStETH;
+        ethToSend = _applyDiscount(ethToSend, discount.discountFactor);
+
+        lockedEtherAmount -= ethToSend;
+
+        _sendValue(request.recipient, ethToSend);
+
+        emit WithdrawalClaimed(_requestId, request.recipient, msg.sender);
+    }
+
+    /// @dev creates a new `WithdrawalRequest` in the queue
+    function _enqueue(uint256 _amountOfStETH, uint256 _amountofShares, address _recipient)
+        internal
+        returns (uint256 requestId)
+    {
+        WithdrawalRequest memory lastRequest = queue[lastRequestId];
+
+        uint256 cumulativeShares = lastRequest.cumulativeShares + _amountofShares;
+        uint256 cumulativeStETH = lastRequest.cumulativeStETH + _amountOfStETH;
+
+        lastRequestId++;
+        requestId = lastRequestId;
+        queue[requestId] = WithdrawalRequest(
+            _toUint128(cumulativeStETH),
+            _toUint128(cumulativeShares),
+            payable(_recipient),
+            _toUint64(block.number),
+            false
+        );
+        requestsByRecipient[_recipient].push(requestId);
+
+        emit WithdrawalRequested(requestId, msg.sender, _recipient, _amountOfStETH, _amountofShares);
+    }
+
+    /// @dev Finalize requests from last finalized one up to `_lastRequestIdToFinalize`
+    function _finalize(uint256 _lastRequestIdToFinalize, uint256 _amountofETH) internal {
+        if (_lastRequestIdToFinalize <= lastFinalizedRequestId) revert InvalidRequestId(_lastRequestIdToFinalize);
+        if (_lastRequestIdToFinalize > lastRequestId) revert InvalidRequestId(_lastRequestIdToFinalize);
+
+        uint128 finalizedStETH = queue[lastFinalizedRequestId].cumulativeStETH;
+        uint128 stETHToFinalize = queue[_lastRequestIdToFinalize].cumulativeStETH - finalizedStETH;
+        uint96 discountFactor = _calculateDiscountFactor(stETHToFinalize, _amountofETH);
+
+        _updateDiscountHistory(_lastRequestIdToFinalize, discountFactor);
+
+        lockedEtherAmount += _applyDiscount(stETHToFinalize, discountFactor);
+        lastFinalizedRequestId = _lastRequestIdToFinalize;
+    }
+
+    /// @dev calculates the sum of stETH and shares for all requests in (`_firstId`, `_lastId`]
+    function _batch(uint256 _firstId, uint256 _lastId)
```

Comment:
maybe find a better name

Reply:
it's a third iteration. Should think about it.